doc/specs/#11000 - Marks/Shell-Integration-Marks.md

```diff
@@ -0,0 +1,431 @@
+D | ✅ Done   | The shell can emit different marks to differentiate between prompt, command, and output
+E | 🐣 Crawl  | Clearing the buffer clears marks
+F | 🐣 Crawl  | Marks stay in the same place you'd expect after resizing the buffer.
+G | 🐣 Crawl  | Users can perform an action to select the previous command's output
+H | 🚶 Walk   | The find dialog can display marks on the scrollbar indicating the position of search matches
+I | 🏃‍♂️ Run    | The terminal can display icons in the gutter, with quick actions for that command (re-run, copy output, etc)
+J | 🏃‍♂️ Run    | The terminal can display a faint horizontal separator between commands in the buffer.
+K | 🚀 Sprint | The terminal can "collapse" content between two marks.
+L | 🚀 Sprint | The terminal can display a sticky header above the control which displays the most recent command 
+M | 🚀 Sprint | The user can open a dialog to manually bookmark a line with a custom comment
+
+## Solution Design
+
+### Supported VT sequences
+
+* [x] iTerm2's OSC `SetMark` (in [#12948])
+* [x] FinalTerm prompt markup sequences
+  - [x] **FTCS_PROMPT** was added in [#13163]
+  - [x] The rest in [#14341] 
+* [ ] additionally, VsCode's FinalTerm prompt markup variant, `OSC 663`
+* [ ] [ConEmu's
+  `OSC9;12`](https://conemu.github.io/en/AnsiEscapeCodes.html#ConEmu_specific_OSC)
+* [ ] Any custom OSC we may want to author ourselves.
+
+The FinalTerm prompt sequences are probably the most complicated version of all
+these, so it's important to give these a special callout. Almost all the other
+VT sequences are roughly equivalent to **FTCS_PROMPT**. The xterm.js / VsCode
+version has additional cases, that they ironically added to work around conpty
+not understanding these sequences originally.
+
+#### FinalTerm sequences
+
+The relevant FinalTerm sequences for marking up the prompt are as follows:
+
+* **FTCS_PROMPT**: `OSC 133 ; A ST`
+  - The start of a prompt. Does nothing all on it's own. Internally, this sets a
+    marker in the buffer indicating we started a prompt at the current cursor
+    position, and that marker will be used when we get a **FTCS_COMMAND_START**
+* **FTCS_COMMAND_START**: `OSC 133 ; B ST`
+  - The start of a commandline (READ: the end of the prompt). When it follows a
+    **FTCS_PROMPT**, it creates a mark in the buffer from the location of the
+    **FTCS_PROMPT** to the current cursor position, with the category of
+    `prompt`
+* **FTCS_COMMAND_EXECUTED**: `OSC 133 ; C ST`
+  - The start of the command output / the end of the commandline.
+* **FTCS_COMMAND_FINISHED**: `OSC 133 ; D ; [Ps] ST`
+  - the end of a command.
+
+Same deal for the **FTCS_COMMAND_EXECUTED**/**FTCS_COMMAND_FINISHED** ones.
+**FTCS_COMMAND_EXECUTED** does nothing until we get a **FTCS_COMMAND_FINISHED**,
+and the `[Ps]` parameter determines the category.
+  - `[Ps] == 0`: success
+  - anything else: error
+
+This whole sequence will get turned into a single mark.

```

Comment:
If you feel like spending time on it, I think a single mermaid diagram might explain the above better than the textual explanation itself. This is especially so since I have trouble understanding how these states relate to each other given just the description above. Like when it says `FTCS_COMMAND_FINISHED` = "the end of a command", then I understand that purely from the name "command finished" already, but I fail to understand what that means in relation to the others. Since it's effectively a state machine I felt like a diagram might replace much of the above... I think? 😅

Reply:
Oh you know what, I had a better picture and I don't know why I didn't include it!

![](https://user-images.githubusercontent.com/18356694/170084383-a1a7b0eb-8d6d-4be8-9472-f9a482e72f22.png)

Okay actually, I do see why I didn't include that - just minor text changes that it needs. Lemme make a new version and include it ☺️